primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,445 @@
+	/// by another process `Err(Some(expiration_timestamp))`.
+	pub fn try_lock<'b>(&'b mut self) -> Result<StorageLockGuard<'a, 'b, L>, Option<L>>
+	where
+		'a: 'b,
+	{
+		match self.try_lock_inner(self.deadline) {
+			Ok(_) => Ok(StorageLockGuard::<'a, 'b> { lock: Some(self) }),
+			Err(e) => Err(e),
+		}
+	}
+
+	/// Try grabbing the lock until its expiry is reached.
+	///
+	/// Returns an error if the lock expired before it could be caught.
+	pub fn spin_lock<'b>(&'b mut self) -> StorageLockGuard<'a, 'b, L>
+	where
+		'a: 'b,
+	{
+		loop {
+			// blind attempt on locking
+			let deadline = match self.try_lock_inner(self.deadline) {
+				Ok(_) => return StorageLockGuard::<'a, 'b, L> { lock: Some(self) },
+				Err(Some(other_locks_deadline)) => other_locks_deadline,
+				_ => L::deadline(), // use the default
+			};
+			L::current().snooze(&deadline);
+		}
+	}
+
+	/// Explicitly unlock the lock.
+	pub fn unlock(&mut self) {
+		self.value_ref.clear();
+	}
+}
+
+/// Extension trait for locks based on [`Timestamp`](::sp_core::offchain::Timestamp).
+///
+/// Allows explicity setting the timeout on construction
+/// instead of using the implicit default timeout of
+/// [`STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS`](Self::STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS).
+pub trait TimeLock<'a>: Sized {
+	/// Provide an explicit deadline timestamp at which the locked state of the lock
+	/// becomes stale and may be dismissed by `fn try_lock(..)`, `fn spin_lock(..)` and others.
+	fn with_deadline(key: &'a [u8], lock_deadline: Timestamp) -> Self;
+}
+
+impl<'a> TimeLock<'a> for StorageLock<'a, Timestamp> {
+	fn with_deadline(key: &'a [u8], lock_deadline: Timestamp) -> Self {
+		Self {
+			value_ref: StorageValueRef::<'a>::persistent(key),
+			deadline: lock_deadline,
+		}
+	}
+}
+
+/// Bound for block numbers which commonly will be implemented by the `frame_system::Trait::BlockNumber`.
+///
+/// This trait has no intrinsic meaning and exists only to decouple `frame_system`
+/// from `runtime` crate and avoid a circular dependency.
+pub trait BlockNumberTrait: Codec + Copy + Clone + Ord + Eq {
+	/// Returns the current block number.
+	///
+	/// Commonly this will be implemented as
+	/// ```ignore
+	/// fn current() -> Self {
+	///     frame_system::Module<Trait>::block_number()
+	/// }
+	/// ```
+	/// but note that the definition of current is
+	/// application specific.
+	fn current() -> Self;
+}
+
+/// Extension for lor storage locks which are based on blocks in addition to a timestamp.
+trait BlockAndTimeLock<'a, B>: Sized
+where
+	B: BlockNumberTrait,
+{
+	fn with_block_and_time_deadline<'k>(
+		key: &'k [u8],
+		block_deadline: B,
+		time_deadline: Timestamp,
+	) -> Self
+	where
+		'k: 'a;
+}
+
+impl<'a, B> BlockAndTimeLock<'a, B> for StorageLock<'a, BlockAndTime<B>>
+where
+	B: BlockNumberTrait,
+{
+	fn with_block_and_time_deadline<'k>(
+		key: &'k [u8],
+		block_deadline: B,
+		time_deadline: Timestamp,
+	) -> Self
+	where
+		'k: 'a,
```

Comment:
```suggestion
	fn with_block_and_time_deadline(
		key: &'a [u8],
		block_deadline: B,
		time_deadline: Timestamp,
	) -> Self
```

again `'k` generic argument does not produce any new lifetime and is actually `'a`